src/merkle/index.rs

```diff
@@ -0,0 +1,160 @@
+use super::Word;
+
+// NODE INDEX
+// ================================================================================================
+
+/// A Merkle tree address to an arbitrary node.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub struct NodeIndex {
+    depth: u8,
+    index: u64,
+}
+
+impl NodeIndex {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new node index.
+    pub const fn new(depth: u8, index: u64) -> Self {
+        Self { depth, index }
+    }
+
+    /// Creates a new node index pointing to the root of the tree.
+    pub const fn root() -> Self {
+        Self { depth: 0, index: 0 }
+    }
+
+    /// Mutates the instance and returns it, replacing the depth.
+    pub const fn with_depth(mut self, depth: u8) -> Self {
+        self.depth = depth;
+        self
+    }
+
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Builds a node to be used as input of a hash function when computing a Merkle path.
+    ///
+    /// Will evaluate the parity of the current instance to define the result.
+    pub fn build_node<T>(&self, slf: Word, sibling: Word) -> [T; 2]
+    where
+        Word: Into<T>,
+    {
+        if self.is_right_sibling() {
+            [sibling.into(), slf.into()]
+        } else {
+            [slf.into(), sibling.into()]
+        }
+    }
+
+    /// Returns the depth of the current instance.
+    pub const fn depth(&self) -> u8 {
+        self.depth
+    }
+
+    /// Returns the index of the current depth.
+    pub const fn index(&self) -> u64 {
+        self.index
+    }
+
+    /// Returns the maximum index for the current depth.
+    pub const fn max_index(&self) -> u64 {
+        1 << self.depth as u64
+    }
+
+    /// Returns true if the current index fits the current depth for a binary tree.
+    pub const fn is_index_valid(&self) -> bool {
+        self.index < self.max_index()
+    }
+
+    /// Returns true if the current instance points to a right sibling node.
+    pub const fn is_right_sibling(&self) -> bool {
+        (self.index & 1) == 1
+    }
+
+    /// Returns `true` if the depth is `0`.
+    pub const fn is_root(&self) -> bool {
+        self.depth == 0
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Traverse towards the leaves, incrementing the depth by `1`.
+    pub fn traverse(&mut self, right: bool) -> &mut Self {
+        self.depth = self.depth.saturating_add(1);
+        self.index <<= 1;
+        self.index += right as u64;
+        self
+    }
+
+    /// Traverse one level towards the root, decrementing the depth by `1`.
+    pub fn backtrack(&mut self) -> &mut Self {
+        self.depth = self.depth.saturating_sub(1);
+        self.index >>= 1;
+        self
+    }
```

Comment:
nit: I'd prefer `move_up()` to `backtrack()`.

Reply:
Done!